packages/shared/pkg/storage/storage_google.go

```diff
@@ -217,61 +217,41 @@ func (g *GCPBucketStorageObjectProvider) WriteTo(dst io.Writer) (int64, error) {
 }
 
 func (g *GCPBucketStorageObjectProvider) WriteFromFileSystem(path string) error {
-	upload := func() error {
-		extraArgs := []string{}
-		if g.limiter != nil {
-			uploadLimiter := g.limiter.GCloudUploadLimiter()
-			if uploadLimiter != nil {
-				semaphoreErr := uploadLimiter.Acquire(g.ctx, 1)
-				if semaphoreErr != nil {
-					return fmt.Errorf("failed to acquire semaphore: %w", semaphoreErr)
-				}
-				defer uploadLimiter.Release(1)
-			}
-
-			extraArgs = append(extraArgs, g.limiter.GCloudCmdLimits(path)...)
-		}
-
-		args := []string{"--scope"}
-		args = append(args, extraArgs...)
-		args = append(args, "--",
-			"gcloud",
-			"storage",
-			"cp",
-			"--verbosity=error",
-			path,
-			fmt.Sprintf("gs://%s/%s", g.storage.bucket.BucketName(), g.path),
-		)
-
-		cmd := exec.CommandContext(
-			g.ctx,
-			"systemd-run",
-			args...,
-		)
-
-		output, err := cmd.CombinedOutput()
-		if err != nil {
-			return fmt.Errorf("failed to upload file to GCS: %w\n%s", err, string(output))
-		}
-
-		return nil
+	bucketName := g.storage.bucket.BucketName()
+	objectName := g.path
+	filePath := path
+
+	uploader, err := NewMultipartUploaderWithRetryConfig(
+		g.ctx,
+		bucketName,
+		objectName,
+		DefaultRetryConfig(),
+	)
+	if err != nil {
+		return fmt.Errorf("failed to create multipart uploader: %w", err)
 	}
 
-	var err error
-	for range gcloudMaxRetries {
-		err = upload()
-		if err != nil {
-			// Failed to upload file, retrying.
-			zap.L().Warn("Failed to upload file to GCS, retrying", zap.Error(err), zap.String("path", g.path))
-
-			continue
-		}
+	fileSize, err := os.Stat(filePath)
+	if err != nil {
+		return fmt.Errorf("failed to get file size: %w", err)
+	}
 
-		// Files was successfully uploaded
-		return nil
+	maxConcurrency := g.limiter.GCloudMaxTasks()
+	start := time.Now()
+	if err := uploader.UploadFileInParallel(g.ctx, filePath, maxConcurrency); err != nil {
```

Comment:
it may make sense to introduce some global concurrency (but let's do it in another PR)